testsuite/diem-backend/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+use diem_crypto::ed25519::{Ed25519PrivateKey, Ed25519PublicKey};
+use diem_types::waypoint::Waypoint;
+use reqwest::{Client, Url};
+use diem_types::chain_id::ChainId;
+use std::fmt;
+use std::str::FromStr;
+use diem_client::Client as JsonRpcClient;
+
+#[derive(Clone)]
+enum InstanceBackend {
+    Swarm,
+    K8s,
+}
+
+#[derive(Clone)]
+pub struct Cluster {
+    validator_instances: Vec<Instance>,
+    fullnode_instances: Vec<Instance>,
+    lsr_instances: Vec<Instance>,
+    vault_instances: Vec<Instance>,
+    mint_key_pair: KeyPair<Ed25519PrivateKey, Ed25519PublicKey>,
+    waypoint: Option<Waypoint>,
+    pub chain_id: ChainId,
+}
+
+impl Cluster {
+    pub fn validator_instances(&self) -> &[Instance] {
+        &self.validator_instances
+    }
+
+    pub fn fullnode_instances(&self) -> &[Instance] {
+        &self.fullnode_instances
+    }
+
+    pub fn lsr_instances(&self) -> &[Instance] {
+        &self.lsr_instances
+    }
+
+    pub fn vault_instances(&self) -> &[Instance] {
+        &self.vault_instances
+    }
+
+    pub fn waypoint(&self) -> Option<Waypoint> {
+        self.waypoint
+    }
+
+    pub fn mint_key_pair(&self) -> &KeyPair<Ed25519PrivateKey, Ed25519PublicKey> {
+        &self.mint_key_pair
+    }
+
+    pub fn validator_and_fullnode_instances(&self) -> impl Iterator<Item = &Instance> {
+        self.validator_instances
+            .iter()
+            .chain(self.fullnode_instances.iter())
+    }
+}
+
+#[derive(Clone)]
+pub struct Instance {
+    peer_name: String,
+    ip: String,
+    json_rpc_port: u32,
+    debug_interface_port: Option<u32>,
+    http_client: Client,
+    backend: InstanceBackend,
+}
+
+impl Instance {
+    pub fn peer_name(&self) -> &str {
+        &self.peer_name
+    }
+
+    pub fn debug_interface_port(&self) -> Option<u32> {
+        self.debug_interface_port
+    }
+
+    pub fn http_clinet(&self) -> Client {
+        self.http_client.clone()
+    }
+
+    pub fn ip(&self) -> &str {
+        &self.ip
+    }
+
+    pub fn json_rpc_port(&self) -> u32 {
+        self.json_rpc_port
+    }
+
+    pub fn json_rpc_url(&self) -> Url {
+        Url::from_str(&format!("http://{}:{}/v1", self.ip(), self.json_rpc_port())).expect("Invalid URL.")
+    }
+
+    pub fn json_rpc_client(&self) -> JsonRpcClient {
+        JsonRpcClient::new(self.json_rpc_url().to_string())
+    }
+}
+
+impl fmt::Display for Instance {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}({})", self.peer_name, self.ip)
+    }
+}
+
+impl fmt::Debug for Instance {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+#[async_trait]
+pub trait Swarm {
+    fn as_cluster(peers: Vec<(String, u32, Option<u32>)>, mint_file: &str, chain_id: ChainId, vasp: bool) -> Cluster;
+}
+
+#[async_trait]
+pub trait Node {
+    fn as_instance(peer_name: String, ip: String, ac_port: u32, debug_interface_port: Option<u32>, http_client: Client) -> Instance;
+}
```

Comment:
These traits should probably be structured differently:

```
trait Node {
    fn as_instance(&self) -> Instance<'a>;
}
```

You shouldn't have to pass the arguments to construct the instance, the type that implements Node should know how to construct it.